### PR TITLE
Ignore keydown event for tab during IME composition

### PR DESCRIPTION
**Pull request by htoor3**
During IME composition, Tab gets fired twice: once as part of the composition of the character and once without. This leads to unexpected behavior such as skipping a textfield during tab and input mutation. The solution is to ignore the keydown event that's a part of composition.

Fixes https://github.com/flutter/flutter/issues/74813

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: lib/web_ui/lib/src/engine/raw_keyboard.dart [hunk 1]
```diff
@@ -96,6 +96,14 @@ class RawKeyboard {
     final FlutterHtmlKeyboardEvent event = FlutterHtmlKeyboardEvent(domEvent as DomKeyboardEvent);
     final String timerKey = event.code!;
 
+    // During IME composition, Tab fires twice (once for composition and once
+    // for regular tabbing behavior), which causes issues. Intercepting the
+    // tab keydown event during composition prevents these issues from occurring.
+    // https://developer.mozilla.org/en-US/docs/Web/API/Element/keydown_event#ignoring_keydown_during_ime_composition
+    if (event.type == 'keydown' && event.key == 'Tab' && event.isComposing) {
+      return;
+    }
```
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/raw_keyboard.dart:105]: I suggest putting this logic in a function:
```suggestion
    if (_shouldIgnore(event)) {
      return;
    }
```
Then add:
```dart
bool _shouldIgnore(FlutterHtmlKeyboardEvent event) {
  // During IME composition, Tab fires twice (once for composition and once
  // for regular tabbing behavior), which causes issues. Intercepting the
  // tab keydown event during composition prevents these issues from occurring.
  // https://developer.mozilla.org/en-US/docs/Web/API/Element/keydown_event#ignoring_keydown_during_ime_composition
  return event.type == 'keydown' && event.key == 'Tab' && event.isComposing;
}
```

File: lib/web_ui/lib/src/engine/raw_keyboard.dart [hunk 2]
```diff
@@ -96,6 +96,14 @@ class RawKeyboard {
     final FlutterHtmlKeyboardEvent event = FlutterHtmlKeyboardEvent(domEvent as DomKeyboardEvent);
     final String timerKey = event.code!;
 
+    // During IME composition, Tab fires twice (once for composition and once
+    // for regular tabbing behavior), which causes issues. Intercepting the
+    // tab keydown event during composition prevents these issues from occurring.
+    // https://developer.mozilla.org/en-US/docs/Web/API/Element/keydown_event#ignoring_keydown_during_ime_composition
+    if (event.type == 'keydown' && event.key == 'Tab' && event.isComposing) {
```
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/raw_keyboard.dart:103]: Do we also need to ignore the corresponding keyup?

**Review by mdebbar — COMMENTED** [review]
Thanks @htoor3 for fixing this issue!

cc @dkwingsmt in case we need to do something similar for the hardware keyboard.

File: lib/web_ui/lib/src/engine/raw_keyboard.dart (hunk 2 above)
Comment (htoor3) [reply lib/web_ui/lib/src/engine/raw_keyboard.dart:103]: It doesn't seem to affect the behavior whether I ignore it or not.  All of the conditional logic in in the keydown/keyup callback `_handleHtmlEvent` just seems to be handling for when the event type is `keydown`

Should I add the logic to ignore it regardless?

Comment (mdebbar) [reply lib/web_ui/lib/src/engine/raw_keyboard.dart:103]: If you tried it with an app, and the framework doesn't complain about it, then it should be fine.

File: lib/web_ui/lib/src/engine/raw_keyboard.dart [hunk 3]
```diff
@@ -88,6 +88,14 @@ class RawKeyboard {
     return _onMacOs;
   }
 
+  // During IME composition, Tab fires twice (once for composition and once
+  // for regular tabbing behavior), which causes issues. Intercepting the
+  // tab keydown event during composition prevents these issues from occurring.
+  // https://developer.mozilla.org/en-US/docs/Web/API/Element/keydown_event#ignoring_keydown_during_ime_composition
```
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/raw_keyboard.dart:94]: This comment describes an implementation detail of the function. I think it should go inside the function. What do you think?

**Review by mdebbar — APPROVED** [review]
LGTM!

File: lib/web_ui/lib/src/engine/raw_keyboard.dart (hunk 3 above)
Comment (htoor3) [reply lib/web_ui/lib/src/engine/raw_keyboard.dart:94]: That definitely makes sense, I was just following the pattern of the `_shouldDoKeyGuard()` function where the comment is included right above the function.  I don't have a strong preference either way

**Pull request merged**